Improve error logging after calling some BIO_ functions

giox069 opened the pull request:
Not really a fix, only a logging improvement: when a BIO_* function fails (ie: socket timeout), nothing is currently logged by freerdp.
This PR will add logging of these failures.
Please note that the code from malloc() to free() is completely untested, because I'm not really able to generate a SSL error, not a system error, for testing.

hardening commented on libfreerdp/core/transport.c:
```diff
@@ -371,6 +371,37 @@ BOOL transport_accept_nla(rdpTransport* transport)
 	return TRUE;
 }
 
+void transport_bio_error_log(char *biofunc, rdpTransport* transport)
```
In this file could you do some early returns so that we don't have these cascading if() ?

giox069 replied:
Done.
Should I squash the two commits into a single one ?

akallabeth replied:
@giox069 ```WLog_*``` functions also record where they were called from. This wrapper function negates this feature, maybe it would be a good idea to add ```__FUNCTION__, __LINE__ and __FILE__``` as arguments and use ```WLog_PrintMessage``` directly.

hardening commented:
LGTM

akallabeth commented on libfreerdp/core/transport.c:
```diff
@@ -371,6 +371,35 @@ BOOL transport_accept_nla(rdpTransport* transport)
+	{
+		while((sslerr = ERR_get_error()))
+		{
+			ERR_error_string(sslerr, buf);
```
@giox069 Maybe it would be a good idea to use ```ERR_error_string_n```

hardening replied:
@akallabeth I have watched the doc and it says that a buffer of at least 120 bytes is enough for ERR_error_string.

akallabeth replied:
@hardening Read that one too, just do not like magic numbers ;)

giox069 replied:
I will add _n and 120 to accomodate @akallabeth paranoid programming style :)

I will also check how to move original `__FUNCTION__`, `__LINE__` and `__FILE__` values inside my function (not yet understood how).

I hope I will be able to do it by tomorrow.

akallabeth commented:
other than my remarks LGTM

giox069 commented:
Added _n, fixed a missing %s on a format string, squashed.

Now the part from malloc() and free() is tested: just try to enter bad credentials when connecting to a microsoft server and SSL will fail (this is expected):
```
[19:55:32:495] [29343:29344] [ERROR][com.freerdp.core] - freerdp_set_last_error ERRCONNECT_AUTHENTICATION_FAILED [0x20009]
[19:55:32:495] [29343:29344] [ERROR][com.freerdp.core.transport] - biofunc BIO_read error: error:14094419:SSL routines:ssl3_read_bytes:tlsv1 alert access denied
[19:55:32:495] [29343:29344] [ERROR][com.freerdp.client.x11] - Freerdp connect error exit status 1
```
The second line comes from my new function.

@akallabeth: I can't find any reference to `__FUNCTION__`, `__LINE__` and `__FILE__` in WLog_ERR code.
```
#define WLog_ERR(tag, fmt, ...) WLog_Print(WLog_Get(tag), WLOG_ERROR, fmt, ## __VA_ARGS__)
```
It seems that WLog_* functions only make use of TAG, which is correctly defined one time per file. In case of trasport.c:
```
#define TAG FREERDP_TAG("core.transport")
```

akallabeth commented:
@giox069 Have a look at the define of ```WLog_Print```,

giox069 commented:
Just updated PR with an improved solution with a macro.
How can I the full log output to the commandline with xfreerdp ? I can't see any option for logging with `xfreerdp /?`

akallabeth commented:
@giox069 Currently there is no command line option (maybe a good idea), you have to set the variable ```WLOG_LEVEL=trace``` and / or to filter some stuff ```WLOG_FILTER=<some tag>:<level>,...```

akallabeth commented:
Looks great now. +1
Any objections to merge?

bmiklautz commented:
@giox069 nice! Thank you.
